iot_verifier/src/gateway_cache.rs

```diff
@@ -1,93 +1,37 @@
-use futures::stream::StreamExt;
+use crate::gateway_updater::MessageReceiver;
 use helium_crypto::PublicKeyBinary;
-use iot_config::{
-    client::{Client as IotConfigClient, ClientError as IotConfigClientError},
-    gateway_info::{GatewayInfo, GatewayInfoResolver},
-};
-use retainer::Cache;
-use std::{sync::Arc, time::Duration};
-use tokio::task::JoinHandle;
-
-/// how long each cached items takes to expire ( 12 hours in seconds)
-const CACHE_TTL: Duration = Duration::from_secs(12 * (60 * 60));
-/// how often to evict expired items from the cache ( every 1 hour)
-const CACHE_EVICTION_FREQUENCY: Duration = Duration::from_secs(60 * 60);
+use iot_config::gateway_info::GatewayInfo;
 
 pub struct GatewayCache {
-    pub iot_config_client: IotConfigClient,
-    pub cache: Arc<Cache<PublicKeyBinary, GatewayInfo>>,
-    pub cache_monitor: JoinHandle<()>,
+    gateway_cache_receiver: MessageReceiver,
 }
 
 #[derive(Debug, thiserror::Error)]
 pub enum GatewayCacheError {
     #[error("gateway not found: {0}")]
     GatewayNotFound(PublicKeyBinary),
-    #[error("error querying iot config service")]
-    IotConfigClient(#[from] IotConfigClientError),
 }
 
 impl GatewayCache {
-    pub fn from_settings(iot_config_client: IotConfigClient) -> Self {
-        let cache = Arc::new(Cache::<PublicKeyBinary, GatewayInfo>::new());
-        let clone = cache.clone();
-        // monitor cache to handle evictions
-        let cache_monitor =
-            tokio::spawn(async move { clone.monitor(4, 0.25, CACHE_EVICTION_FREQUENCY).await });
+    pub fn new(gateway_cache_receiver: MessageReceiver) -> Self {
         Self {
-            iot_config_client,
-            cache,
-            cache_monitor,
-        }
-    }
-
-    pub async fn prewarm(&self) -> anyhow::Result<()> {
-        tracing::info!("starting prewarming gateway cache");
-        let mut gw_stream = self
-            .iot_config_client
-            .clone()
-            .stream_gateways_info()
-            .await?;
-        while let Some(gateway_info) = gw_stream.next().await {
-            _ = self.insert(gateway_info).await;
+            gateway_cache_receiver,
         }
-        tracing::info!("completed prewarming gateway cache");
-        Ok(())
     }
 
     pub async fn resolve_gateway_info(
         &self,
         address: &PublicKeyBinary,
     ) -> Result<GatewayInfo, GatewayCacheError> {
-        match self.cache.get(address).await {
+        match self.gateway_cache_receiver.borrow().get(address) {
             Some(hit) => {
-                metrics::increment_counter!("oracles_iot_verifier_gateway_cache_hit");
-                Ok(hit.value().clone())
+                metrics::increment_counter!("oracles_iot_verifier_gateway_found");
```

Comment:
i know we're no longer falling back to checking directly with the config service when this call fails to return a result but i think it would still be helpful to label these metrics to highlight that we're hitting a cache, not the direct source of data

Reply:
fixed here: 13fc436c2496633a4f744bd7c48299df873fc97d